File: reminder.py
from datetime import datetime

class Reminder:
    def __init__(self, subscription, days_before):
        self.subscription = subscription
        self.days_before = days_before

    def send_reminder(self):
        remaining_days = (self.subscription.expiry_date - datetime.now()).days
        if remaining_days <= self.days_before:
            return f"Reminder: Your {self.subscription.software_name} license expires in {remaining_days} days."
        return None

File: java-version/Reminder.java
// File: Reminder.java
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

public class Reminder {
    private Subscription subscription;
    private int daysBefore;

    public Reminder(Subscription subscription, int daysBefore) {
        this.subscription = subscription;
        this.daysBefore = daysBefore;
    }

    public void sendReminder() {
        long daysLeft = ChronoUnit.DAYS.between(LocalDate.now(), subscription.getExpiryDate());
        if (daysLeft <= daysBefore) {
            System.out.println("Reminder: Your " + subscription.getSoftwareName() + " license will expire in " + daysLeft + " days.");
        }
    }
}
